posts/semilinear-maps.md

```diff
@@ -0,0 +1,22 @@
+
+This left the more arduous option, namely to redefine `linear_map` to also include semilinear maps. A semilinear map `f` is a map from an `R`-module to an `S`-module with a ring homomorphism `σ` between `R` and `S`, such that `f (c • x) = (σ c) • (f x)`. If we plug in the identity into `σ`, we get regular linear maps, and if we plug in the complex conjugate, we get conjugate linear maps. There are also other examples (e.g. Frobenius-linear maps) where this is useful which are covered by this general formulation. This implied a major refactor: we had to replace the basic definition of `R`-linear maps `E →ₗ[R] F` by `σ`-semilinear maps `E →ₛₗ[σ] F` while keeping the original notation for plain linear maps, and deal with the various problems that this inevitably created further down the import tree. The same also had to be done for linear equivalences, continuous linear maps/equivalences, and linear isometries. This idea had first been proposed by Yury Kudryashov [about a year ago](https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/4770.20smul_comm_class/near/214442983), but it then took several months to build up the motivation to embark on this project. Last August, Heather Macbeth, Rob Lewis and I finally managed to start working on it, and the result was [merged](https://github.com/leanprover-community/mathlib/pull/9272) into mathlib in late September.
+
+The main issue that we had to overcome involved composition of semilinear maps, and `symm` for linear equivalences. Suppose we have `f : E₁ →ₛₗ[σ₁₂] E₂` and `g : E₂ →ₛₗ[σ₂₃] E₃`, we would naturally end up with `g.comp f : E₁ →ₛₗ[σ₂₃.comp σ₁₂] E₃`. This is undesirable in most cases: for example, we want the composition of two conjugate-linear maps to be a regular linear map, not a `conj.comp conj`-linear map. Likewise, if `e : E →ₛₗ[conj] F`, we don't want `e.symm` to have type `E →ₛₗ[conj.symm] F`! To solve this issue, we created two typeclasses to make Lean infer the right ring homomorphism. The first one is `[ring_hom_comp_triple σ₁₂ σ₂₃ σ₁₃]` which expresses the fact that `σ₂₃.comp σ₁₂ = σ₁₃`, and the second one is `[ring_hom_inv_pair σ₁₂ σ₂₁]` which states that `σ₁₂` and `σ₂₁` are inverses of each other. The ring homomorphism `σ₁₃` (resp. `σ₂₁`) is inferred silently by the typeclass system using `out_param`. There is also a third typeclass `[ring_hom_surjective σ]`, which is a necessary assumption to generalize some basic lemmas.
```

Comment:
I think it would be nice to include a couple of example statements and use to make this easier to understand.

Reply:
I have now added some simplified examples to make this point clearer. Also I added a footnote using what the Internet told me was correct Markdown syntax, but I haven't test it.

Reply:
I still think it would be more concrete to give the full type of `linear_map.comp` for instance (or a simpler example if you can find one).